File: services/server/src/main/java/com/yaytsa/server/infrastructure/persistence/entity/TasteProfileEntity.java
package com.yaytsa.server.infrastructure.persistence.entity;

import jakarta.persistence.*;
import java.time.OffsetDateTime;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.Array;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "taste_profile")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class TasteProfileEntity {

  @Id
  @Column(name = "user_id")
  private UUID userId;

  @OneToOne(fetch = FetchType.LAZY)
  @MapsId
  @JoinColumn(name = "user_id", foreignKey = @ForeignKey(name = "fk_taste_profile_user"))
  private UserEntity user;

  @JdbcTypeCode(SqlTypes.JSON)
  @Column(columnDefinition = "jsonb", nullable = false)
  private String profile = "{}";

  @Column(name = "summary_text", columnDefinition = "TEXT")
  private String summaryText;

  @JdbcTypeCode(SqlTypes.VECTOR)
  @Array(length = 768)
  @Column(name = "embedding_mert", columnDefinition = "vector(768)")
  private float[] embeddingMert;

  @JdbcTypeCode(SqlTypes.VECTOR)
  @Array(length = 512)
  @Column(name = "embedding_clap", columnDefinition = "vector(512)")
  private float[] embeddingClap;

  @Column(name = "track_count", nullable = false)
  private int trackCount;

  @Column(name = "rebuilt_at")
  private OffsetDateTime rebuiltAt;
}

File: infra-persistence/ml/src/main/kotlin/dev/yaytsa/persistence/ml/entity/TasteProfileEntity.kt
package dev.yaytsa.persistence.ml.entity

import dev.yaytsa.persistence.shared.converter.FloatArrayAttributeConverter
import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.Instant
import java.util.UUID

@Entity
@Table(name = "taste_profiles", schema = "core_v2_ml")
class TasteProfileEntity(
    @Id
    @Column(name = "user_id")
    val userId: UUID = UUID.randomUUID(),
    @Column(columnDefinition = "TEXT")
    val profile: String? = null,
    @Column(name = "summary_text")
    val summaryText: String? = null,
    @Column(name = "rebuilt_at", nullable = false)
    val rebuiltAt: Instant = Instant.now(),
    @Column(name = "track_count", nullable = false)
    val trackCount: Int = 0,
    @Convert(converter = FloatArrayAttributeConverter::class)
    @Column(name = "embedding_mert", columnDefinition = "float[]")
    val embeddingMert: FloatArray? = null,
    @Convert(converter = FloatArrayAttributeConverter::class)
    @Column(name = "embedding_clap", columnDefinition = "float[]")
    val embeddingClap: FloatArray? = null,
)
